checker/src/org/checkerframework/checker/index/searchindex/SearchIndexAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,194 @@
+            }
+            if (AnnotationUtils.areSame(a2, BOTTOM)) {
+                return a1;
+            }
+            if (isSubtype(a1, a2)) {
+                return a2;
+            }
+            if (isSubtype(a2, a1)) {
+                return a1;
+            }
+            // If neither is a subtype of the other, then create an
+            // annotation that includes only their overlapping values.
+
+            // Each annotation is either NegativeIndexFor or SearchIndexFor.
+            List<String> arrayIntersection = IndexUtil.getValueOfAnnotationWithStringArgument(a1);
+            arrayIntersection.retainAll(IndexUtil.getValueOfAnnotationWithStringArgument(a2));
+
+            if (arrayIntersection.size() == 0) {
+                return UNKNOWN;
+            }
+
+            if (AnnotationUtils.areSameByClass(a1, SearchIndexFor.class)
+                    || AnnotationUtils.areSameByClass(a2, SearchIndexFor.class)) {
+                return createSearchIndexFor(
+                        Arrays.asList(arrayIntersection.toArray(new String[0])));
+            } else {
+                return createNegativeIndexFor(
+                        Arrays.asList(arrayIntersection.toArray(new String[0])));
+            }
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexUnknown.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexBottom.class)) {
+                return true;
+            }
+            if (AnnotationUtils.areSameByClass(subAnno, SearchIndexUnknown.class)) {
+                return false;
+            }
+            if (AnnotationUtils.areSameByClass(superAnno, SearchIndexBottom.class)) {
+                return false;
+            }
+
+            // Each annotation is either NegativeIndexFor or SearchIndexFor.
+            List<String> superArrays = IndexUtil.getValueOfAnnotationWithStringArgument(superAnno);
+            List<String> subArrays = IndexUtil.getValueOfAnnotationWithStringArgument(subAnno);
+
+            // Subtyping requires:
+            //  * subtype is NegativeIndexFor or supertype is SearchIndexFor
+            //  * subtype's arrays are a superset of supertype's arrays
+            return ((AnnotationUtils.areSameByClass(subAnno, NegativeIndexFor.class)
+                            || AnnotationUtils.areSameByClass(superAnno, SearchIndexFor.class))
+                    && subArrays.containsAll(superArrays));
+        }
+    }
+
+    /** Create a new {@code @NegativeIndexFor} annotation with the given arrays as its arguments. */
+    AnnotationMirror createNegativeIndexFor(List<String> arrays) {
+        if (arrays.size() == 0) {
+            return UNKNOWN;
+        }
+        AnnotationBuilder builder = new AnnotationBuilder(processingEnv, NegativeIndexFor.class);
```

Comment:
Sort and remove duplicates? (Sorting doesn't matter for correctness, but will make error messages easier to read.)